packages/widget-v2/src/widget/TokenAndChainSelectorFlow/TokenAndChainSelectorFlow.tsx

```diff
@@ -0,0 +1,110 @@
+import NiceModal, { useModal } from '@ebay/nice-modal-react';
+import { Modal, ModalProps } from '../../components/Modal';
+import { Column } from '../../components/Layout';
+import { styled } from 'styled-components';
+import { useAtom } from 'jotai';
+import { ChainWithAsset, ClientAsset, skipAssets } from '../../state/skip';
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import { VirtualList } from '../../components/VirtualList';
+import { RowItem, Skeleton } from './TokenAndChainSelectorFlowRowItem';
+import { SearchInput } from './TokenAndChainSelectorFlowSearchInput';
+import { hashObject } from '../../utils/misc';
+
+export type TokenAndChainSelectorFlowProps = ModalProps & {
+  onSelect: (token: ClientAsset | null) => void;
+  chainsContainingAsset?: ChainWithAsset[];
+  asset?: ClientAsset;
+};
+
+export const TokenAndChainSelectorFlow = NiceModal.create(
+  (modalProps: TokenAndChainSelectorFlowProps) => {
+    const modal = useModal();
+    const { onSelect, chainsContainingAsset, asset } = modalProps;
+    const [{ data: assets, isPending }] = useAtom(skipAssets);
+
+    const [showSkeleton, setShowSkeleton] = useState(true);
+    const [searchQuery, setSearchQuery] = useState<string>('');
+
+    const filteredAssets = useMemo(() => {
+      if (!assets) return;
+      const filtered = assets.filter((asset) =>
+        asset.symbol?.toLowerCase().includes(searchQuery.toLowerCase())
+      );
+      return filtered;
+    }, [searchQuery, assets]);
+
+    const filteredChains = useMemo(() => {
+      if (!chainsContainingAsset) return;
+      const filtered = chainsContainingAsset.filter(
+        (chain) =>
+          chain.chain_name?.toLowerCase().includes(searchQuery.toLowerCase()) ||
+          chain.pretty_name?.toLowerCase().includes(searchQuery.toLowerCase())
+      );
+      return filtered;
+    }, [searchQuery, chainsContainingAsset]);
+
+    useEffect(() => {
+      if (!isPending && assets) {
+        const timer = setTimeout(() => {
+          setShowSkeleton(false);
+        }, 100);
+        return () => clearTimeout(timer);
+      }
+    }, [isPending, assets]);
```

Comment:
I added a skeleton/fallback on mount because even when the data exists, it takes a while for it to fully render which would normally cause a delay before the modal opens. This way we open the modal faster and swap out the real List afterwards and it doesn't feel as laggy.